core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -403,13 +533,20 @@ class LogManager(val logDirs: Array[File],
   def getLog(topicPartition: TopicPartition): Option[Log] = Option(logs.get(topicPartition))
 
   /**
-   * Create a log for the given topic and the given partition
    * If the log already exists, just return a copy of the existing log
+   * Otherwise if isNew=true or if there is no offline log directory, create a log for the given topic and the given partition
+   * Otherwise throw KafkaStorageException
+   *
+   * @param isNew Whether the replica should have existed on the broker or not
+   * @throws KafkaStorageException if isNew=false, log is not found in the cache and there is offline log directory on the broker
    */
-  def createLog(topicPartition: TopicPartition, config: LogConfig): Log = {
+  def getOrCreateLog(topicPartition: TopicPartition, config: LogConfig, isNew: Boolean = false): Log = {
     logCreationOrDeletionLock synchronized {
-      // create the log if it has not already been created in another thread
       getLog(topicPartition).getOrElse {
+        // create the log if it has not already been created in another thread
+        if (!isNew && offlineLogDirs.nonEmpty)
+          throw new KafkaStorageException(s"Can not create log for $topicPartition because log directories ${offlineLogDirs.mkString(",")} are offline")
```

Comment:
My previous convention is that:
1) IOException should maybeAddLogDirFailureEvent.
2) IOException should either be logged or be re-thrown as KafkaStorageException.
3) KafkaStorageException should not trigger maybeAddLogDirFailureEvent.

Previously I feel it is simpler to catch IOException in the ReplicaManager similar to how we handle most other exceptions that happen during request processing. And I intend to avoid adding big try/catch because it typically makes code review harder.

Now I agree with you that it is a good idea to keep it consistent and try to catch IOException before the ReplicaManager. I have updated the patch so that the ReplicaManger now only catches IOException while writing to checkpoint file. Most other IOException is now caught in Log, LogManager and CheckpointFile. These IOException will trigger maybeAddLogDirFailureEvent() and be re-thrown as KafkStorageException. I think this address the issue here.